crates/warg/src/package/validate.rs

```diff
@@ -191,267 +251,262 @@ impl ValidationState {
         &self,
         record: &model::PackageRecord,
     ) -> Result<(), ValidationError> {
-        if let ValidationState::Initialized(state) = &self {
-            if record.timestamp < state.last_timestamp {
+        if let Some(root) = &self.root {
+            if record.timestamp < root.timestamp {
                 return Err(ValidationError::TimestampLowerThanPrevious);
             }
         }
+
         Ok(())
     }
 
     fn validate_record_entries(
-        self,
-        key_id: signing::KeyID,
-        record: &model::PackageRecord,
-    ) -> Result<EntryValidationState, ValidationError> {
-        let mut entry_validation_state = match self {
-            ValidationState::Uninitialized => None,
-            ValidationState::Initialized(state) => Some(state.entry_state),
-        };
+        &mut self,
+        signer_key_id: &signing::KeyID,
+        timestamp: SystemTime,
+        entries: &[model::PackageEntry],
+    ) -> Result<(), ValidationError> {
+        for entry in entries {
+            if let Some(permission) = entry.required_permission() {
+                self.check_key_permission(signer_key_id, permission)?;
+            }
 
-        for entry in &record.entries {
-            entry_validation_state = match entry_validation_state {
-                Some(state) => Some(state.validate_next(key_id.clone(), entry)?),
-                None => Some(EntryValidationState::validate_first(key_id.clone(), entry)?),
-            };
-        }
+            // Process an init entry specially
+            if let model::PackageEntry::Init {
+                hash_algorithm,
+                key,
+            } = entry
+            {
+                self.validate_init_entry(signer_key_id, *hash_algorithm, key)?;
+                continue;
+            }
 
-        match entry_validation_state {
-            Some(state) => Ok(state),
-            None => Err(ValidationError::InitialRecordDoesNotInit),
+            // Must have seen an init entry by now
+            if self.algorithm.is_none() {
```

Comment:
Consider adding `initialized(&self) -> bool` to use places where we want to know if initialized to make it more explicit.